File: src/ui/webview/note_webview.cpp
/*
 * This file is part of the EverSticky project (https://eversticky.joeeey.com).
 * Copyright (c) 2021 Joey Miller.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include "note_webview.h"

#include <iostream>

#include <QAbstractTextDocumentLayout>
#include <QApplication>
#include <QDropEvent>
#include <QScrollBar>
#include <QTextBlock>
#include <QThread>
#include <QWebChannel>
#include <QWebEngineSettings>

#include "js_interface.h"
#include "note_formatter.h"
#include "ui/note_widget.h"


NoteWebview::NoteWebview(QWidget *context) : QWebEngineView(context)
{
    setObjectName("NoteWebview");

    setFixedHeight(200);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);

    page()->setBackgroundColor(Qt::transparent);
    settings()->setAttribute(QWebEngineSettings::ShowScrollBars, false);

    // Create channel to communicate with the webview
    channel = new QWebChannel(this);
    this->page()->setWebChannel(channel);
    jsInterface = new JsInterface(this);
    channel->registerObject(QString("parentWebEngine"), jsInterface);

    // When webview javascript indicates page is loaded, emit another signal.
    // This signal indicates the webview is ready to be sent content to display (via setText())
    connect(jsInterface, &JsInterface::pageLoaded, this, &NoteWebview::initialNoteLoad);

    connect(jsInterface, &JsInterface::pageHeightUpdated, this, &NoteWebview::updatePageHeight);
    connect(jsInterface, &JsInterface::textUpdated, this, &NoteWebview::textUpdated);
    connect(jsInterface, &JsInterface::ensureWebviewCaretVisible, this, &NoteWebview::ensureWebviewCaretVisible);

    installEventFilter(this);

    openBlankPage();
}

void NoteWebview::updatePageHeight(int newHeight)
{
    int pageBottomPadding = 16;
    this->setFixedHeight(newHeight + pageBottomPadding);
}

void NoteWebview::openBlankPage()
{
    QFile apiFile(":/qtwebchannel/qwebchannel.js");
    if(!apiFile.open(QIODevice::ReadOnly))
        qCritical() << "Error loading qwebchannel.js to inject into webview...";
    QString apiScript = QString::fromLatin1(apiFile.readAll());
    apiFile.close();
    this->page()->runJavaScript(apiScript);

    // Load note js resource
    QFile jsFile(":/ui/webview/note_webview.js");
    if(!jsFile.open(QIODevice::ReadOnly))
        qCritical() << "Error loading note_webview.js to inject into webview...";
    QString jsScript = QString::fromLatin1(jsFile.readAll());
    this->page()->runJavaScript(jsScript);

    // Load note keyhandler js resource
    QFile jsKeyHandlerFile(":/ui/webview/note_webview_keyhandler.js");
    if(!jsKeyHandlerFile.open(QIODevice::ReadOnly))
        qCritical() << "Error loading note_webview_keyhandler.js to inject into webview...";
    QString jsKeyHandlerScript = QString::fromLatin1(jsKeyHandlerFile.readAll());

    // Load note css resource
    QFile cssFile(":/ui/webview/note_webview.css");
    if(!cssFile.open(QIODevice::ReadOnly))
        qCritical() << "Error loading note_webview.css to inject into webview...";
    QString noteCss = QString::fromLatin1(cssFile.readAll());

    setHtml("\
        <html> \
            <head>\
                <script>" + apiScript + "</script> \
                <script>" + jsScript + "</script> \
                <script>" + jsKeyHandlerScript + "</script> \
                <style>" + noteCss + "</style> \
            </head> \
            <body></body>\
        </html>"
    );
}

// Don't call until first page load is finished.
void NoteWebview::setText(QString text)
{
    jsInterface->setInnerHtml(text);
}

bool NoteWebview::eventFilter(QObject*, QEvent *event)
{
    // Would have rather used wheelEvent(), but on some Qt versions (such as Qt 5.12.8 default on Ubuntu 20.04)
    // wheel events aren't caught by the QWebEngineView itself. Need to attach an event filter to child QWidgets that appear
    // during page load.
    // https://bugreports.qt.io/browse/QTBUG-43602

    if (event->type() == QEvent::ChildAdded)
    {
        // Attach event filter to all children as they appear
        QChildEvent *e = static_cast<QChildEvent *>(event);
        QObject *child = e->child();
        child->installEventFilter(this);

        return true;
    }

    if (event->type() == QEvent::Wheel)
    {
        // Consume wheel event so it can be handled by NoteScrollArea
        return true;
    }

    return false;
}

File: src/ui/webview/note_webview.js
/*
 * This file is part of the eversticky project (https://eversticky.joeeey.com).
 * Copyright (c) 2021 Joey Miller.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

var content;
var parentWebEngine;

var prevCaretY;
var prevCaretYStart;
var prevCaretYEnd;

window.onload = function() {
    main();
}

function main()
{
    // Allow content of webview to be user-editable
    document.documentElement.contentEditable = true;

    const webChannel = new QWebChannel(qt.webChannelTransport, function (channel) {
        parentWebEngine = channel.objects.parentWebEngine;

        // Declare note observers
        var observer = new MutationObserver(function()
        {
            domChanged();
        });
        //
        var resizeObserver = new ResizeObserver(function()
        {
            let body = document.body;
            let bodyStyle = window.getComputedStyle(body);
            let bodyMarginTop = parseInt(bodyStyle.getPropertyValue('margin-top'));
            let bodyMarginBottom = parseInt(bodyStyle.getPropertyValue('margin-bottom'));

            let newDomHeight = content.offsetHeight + bodyMarginTop + bodyMarginBottom;
            parentWebEngine.domHeightResized(newDomHeight);
        });


        // Handler for signal updateInnerHtml() from JsInterface
        parentWebEngine.updateInnerHtml.connect(function(text)
        {
            // Replace body with new note
            document.body.innerHTML = text;

            // Connect observers to new note container div
            content = document.getElementById('en-note');
            observer.observe(content, {attributes: true, characterData: true, childList: true, subtree: true});
            resizeObserver.observe(content);

            // Watch for selection changes
            document.addEventListener("selectionchange", selectionChangeEvent);


            /*
             *  Usability enhancements
             */

            // Connect event listener to all checkboxes to ensure 'clicked' attribute
            // indicates the state of the checkbox.
            let checkboxes = document.querySelectorAll("input.en-todo");
            for(var i = 0; i < checkboxes.length; i++) {
                let checkbox = checkboxes[i];
                checkboxAddEventListener(checkbox);
            };

            // Attached to window instead of #en-note to ensure we always
            // get the event (regardless of what's explicitly focussed).
            window.addEventListener("keydown", handleKeyDownEvent);

            // Disable note zoom
            window.addEventListener("wheel", wheelEvent, { passive: false });
        });


        parentWebEngine.insertCheckbox.connect(function(){
            insertCheckbox();
        });
        parentWebEngine.insertDivider.connect(function(){
            insertDivider();
        });
        parentWebEngine.insertBulletedList.connect(function(){
            insertBulletedList();
        });
        parentWebEngine.insertNumberedList.connect(function(){
            insertNumberedList();
        });


        // Notify JsInterface that initial DOM load is complete
        parentWebEngine.domLoaded();
    })
}


/*
 *
 */
function checkboxAddEventListener(checkbox)
{
    checkbox.addEventListener("click", checkboxClickEvent);

    // Don't allow right-clicking on checkboxes
    checkbox.addEventListener("contextmenu", checkboxContextMenuEvent);
}

function checkboxClickEvent(event)
{
    let checkbox = event.currentTarget;
    if(checkbox.checked) {
        checkbox.removeAttribute('checked');
    } else {
        checkbox.setAttribute('checked', 'true');
    }
}

function checkboxContextMenuEvent(event)
{
    event.preventDefault();
    return;
}

function checkCaretYPosChanged(selection)
{
    const selectionRangeStart = selection.getRangeAt(0).cloneRange();
    selectionRangeStart.collapse(true);
    const caretYStart = getCaretPositionAtSelection(selectionRangeStart).top;

    const selectionRangeEnd = selection.getRangeAt(0).cloneRange();
    selectionRangeEnd.collapse(false);
    const caretYEnd = getCaretPositionAtSelection(selectionRangeEnd).top;

    if(caretYStart !== prevCaretYStart) {
        parentWebEngine.domCaretYPosChanged(caretYStart);
    } else if(caretYEnd !== prevCaretYEnd) {
        parentWebEngine.domCaretYPosChanged(caretYEnd);
    }

    prevCaretYStart = caretYStart;
    prevCaretYEnd = caretYEnd;
}

function domChanged()
{
    // XMLSerializer needed to ensure string is valid XML
    // Otherwise, for example, .outerHTML returns <br> instead of <br/>
    parentWebEngine.domChanged(new XMLSerializer().serializeToString(content));
}

function insertCheckbox()
{
    document.execCommand('insertHTML', false, '<input class=\"en-todo\" src=\"data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=\" type=\"image\" />');
}

function insertDivider()
{
    document.execCommand('insertHorizontalRule', false);
}

function insertBulletedList()
{
    document.execCommand('insertUnorderedList', false);
}

function insertNumberedList()
{
    document.execCommand('insertOrderedList', false);
}

function selectionChangeEvent(event)
{
    const selection = window.getSelection();
    checkCaretYPosChanged(selection);
}

function wheelEvent(event)
{
    if(event.ctrlKey)
        event.preventDefault();
        return;
}
